pkg/redirectpolicy/manager.go

```diff
@@ -0,0 +1,648 @@
+		log.WithFields(logrus.Fields{
+			logfields.K8sNamespace:             config.namespace,
+			logfields.LRPName:                  config.name,
+			logfields.K8sSvcID:                 config.serviceID,
+			logfields.LRPSvcPorts:              config.svcPorts,
+			logfields.LRPLocalEndpointSelector: config.backendSelector,
+			logfields.LRPBackendPorts:          config.backendPorts,
+		}).Debug("Add local redirect policy")
+		for _, svcPort := range config.svcPorts {
+			svcAddr := svcCache.GetServiceAddrWithPortAndType(*config.serviceID,
+				lb.SVCTypeClusterIP, svcPort.Port)
+			if svcAddr != nil {
+				rpm.updateConfigSvcFrontend(&config, svcAddr, svcPort.Port)
+			}
+		}
+		pods := rpm.getLocalPodsForPolicy(&config, podStore)
+		if len(pods) == 0 {
+			return true, nil
+		}
+		rpm.upsertForConfigTypeSvc(&config, pods...)
+
+	case lrpConfigTypeNone:
+	default:
+		return false, fmt.Errorf("Invalid local redirect policy config %v", config)
+	}
+
+	return true, nil
+}
+
+// DeleteRedirectPolicy deletes the internal state associated with the given policy.
+func (rpm *Manager) DeleteRedirectPolicy(config LRPConfig) (bool, error) {
+	rpm.mutex.Lock()
+	defer rpm.mutex.Unlock()
+
+	storedConfig := rpm.policyConfigs[config.id]
+	if storedConfig == nil {
+		return false, fmt.Errorf("Local redirect policy to be deleted not found")
+	}
+	log.WithFields(logrus.Fields{"policyID": config.id}).
+		Debug("Delete local redirect policy")
+
+	if storedConfig.lrpType == lrpConfigTypeSvc {
+		rpm.deletePolicyService(*storedConfig.serviceID)
+	} else {
+		rpm.deletePolicyFrontend(storedConfig, storedConfig.frontendMappingForAddr.feAddr)
+	}
+	for p, pp := range rpm.podPolicy {
+		var newPolicyList []podPolicyInfo
+		for _, info := range pp {
+			if info.policyID != storedConfig.id {
+				newPolicyList = append(newPolicyList, info)
+			}
+		}
+		if len(newPolicyList) > 0 {
+			rpm.podPolicy[p] = newPolicyList
+		} else {
+			delete(rpm.podPolicy, p)
+		}
+	}
+	rpm.deletePolicyConfig(storedConfig)
+	return false, nil
+}
+
+// OnAddService handles Kubernetes service add events, and updates the internal state
+// for the policy config associated with the service.
+func (rpm *Manager) OnAddService(svcID k8s.ServiceID, svcCache *k8s.ServiceCache, podStore cache.Store) {
+	rpm.mutex.Lock()
+	defer rpm.mutex.Unlock()
+	if len(rpm.policyConfigs) == 0 {
+		return
+	}
+
+	// Check if this service is selected by any of the current policies.
+	if id, ok := rpm.services[svcID]; ok {
+		config := rpm.policyConfigs[id]
+		if !config.checkNamespace(svcID.Namespace) {
+			return
+		}
+		for _, svcPort := range config.svcPorts {
+			svcAddr := svcCache.GetServiceAddrWithPortAndType(*config.serviceID,
+				lb.SVCTypeClusterIP, svcPort.Port)
+			if svcAddr != nil {
+				rpm.updateConfigSvcFrontend(config, svcAddr, svcPort.Port)
+			}
+		}
+		pods := rpm.getLocalPodsForPolicy(config, podStore)
+		if len(pods) > 0 {
+			rpm.upsertForConfigTypeSvc(config, pods...)
+		}
+	}
```

Comment:
Style nit: We could check the inverse of this condition like so:

```
var id sometype
if id, ok := rpm.services[svcID]; !ok {
  return
}
<rest of code>
```

and that'll allow us to dedent the happy-path.